modules/programs/firejail.nix

```diff
@@ -0,0 +1,106 @@
+      in
+      ''
+        cat <<_EOF >$out/bin/${command}
+        #! ${pkgs.runtimeShell} -e
+        exec ${cfg.firejailBinary} ${args} ${toString opts.executable} "\$@"
+        _EOF
+        chmod 0755 $out/bin/${command}
+
+        ${optionalString (opts.desktop != null) ''
+          substitute ${opts.desktop} $out/share/applications/$(basename ${opts.desktop}) \
+            --replace ${opts.executable} $out/bin/${command}
+        ''}
+      '') cfg.wrappedBinaries)}
+    '';
+
+in
+{
+  options.programs.firejail = {
+    firejailBinary = mkOption {
+      type = types.path;
+      default = "/run/wrappers/bin/firejail";
+      description = mdDoc "The firejail executable with setuid installed using your package manager";
+      example = "/bin/firejail";
+    };
+
+    wrappedBinaries = mkOption {
+      type = types.attrsOf (types.either types.path (types.submodule {
+        options = {
+          executable = mkOption {
+            type = types.path;
+            description = mdDoc "Executable to run sandboxed";
+            example = literalExpression ''"''${getBin pkgs.firefox}/bin/firefox"'';
+          };
+          profile = mkOption {
+            type = types.nullOr types.path;
+            default = null;
+            description = mdDoc "Profile to use";
+            example = literalExpression ''"''${pkgs.firejail}/etc/firejail/firefox.profile"'';
+          };
+          desktop = mkOption {
+            type = types.nullOr types.path;
+            default = null;
+            description = mdDoc "Desktop file of the application to replace the executable in";
+            example = literalExpression ''"''${pkgs.firefox}/share/applications/firefox.profile"'';
+          };
+          extraArgs = mkOption {
+            type = types.listOf types.str;
+            default = [ ];
+            description = mdDoc "Extra arguments to pass to firejail";
+            example = [ "--private=~/.firejail_home" ];
+          };
+        };
+      }));
+      default = { };
+      example = literalExpression ''
+        {
+          firefox = {
+            executable = "''${getBin pkgs.firefox}/bin/firefox";
+            profile = "''${pkgs.firejail}/etc/firejail/firefox.profile";
+            desktop = "''${pkgs.firefox}/share/applications/firefox.desktop";
+            extraArgs = [ "--private" ];
+          };
+          mpv = "''${getBin pkgs.mpv}/bin/mpv";
+        }
+      '';
+      description = mdDoc ''
+        Wrap the binaries in firejail and place them in your path.
+        You still need to install firejail globally using your package manager.
+        You will get file collisions if you put the actual application binary in
+        the environment (such as by adding the application package to
+        `home.packages`), and applications started via .desktop files are not
+        wrapped if they specify the absolute path to the binary.
+      '';
+    };
+  };
+
+  config = mkIf (cfg.wrappedBinaries != { }) {
```

Comment:
Wouldn't it be a good idea to perform an assertion verifying that the path for the firejail binary exists (maybe with the `pathExists` builtin that should work)

Reply:
That would be IFD. If we want an assertion, we could do something like [what the NixOS security wrappers module does](https://github.com/NixOS/nixpkgs/blob/1e2590679d0ed2cee2736e8b80373178d085d263/nixos/modules/security/wrappers/default.nix#L285-L286) to check this at build time.

Reply:
@ncfavier what does IFD mean?

Reply:
Got it, makes sense now. Yes, the idea is that as long the firejailBinary path is verified (currently it isn't), then that would work fine.